internal/ethapi/api.go

```diff
@@ -2606,25 +2606,73 @@ func (api *BlockChainAPI) GetStakerROIMasternode(masternode common.Address) floa
 	return 100.0 / float64(totalCap.Div(totalCap, voterRewardAYear).Uint64())
 }
 
-type currentTotalMinted struct {
-	TotalMinted  *hexutil.Big `json:"totalMinted"`
-	LastEpochNum *hexutil.Big `json:"lastEpochNum"`
-	BlockHash    common.Hash  `json:"blockHash"`
-	BlockNumber  *hexutil.Big `json:"blockNumber"`
-}
-
-func (api *BlockChainAPI) GetCurrentTotalMinted(ctx context.Context) (*currentTotalMinted, error) {
-	statedb, header, err := api.b.StateAndHeaderByNumber(ctx, rpc.LatestBlockNumber)
+type tokenSupply struct {
+	PostUpgradeTotalMinted *hexutil.Big `json:"postUpgradeTotalMinted"`
+	PreUpgradeTotalMinted  *hexutil.Big `json:"preUpgradeTotalMinted"`
+	PostUpgradeTotalBurned *hexutil.Big `json:"postUpgradeTotalBurned"`
+	TotalMinted            *hexutil.Big `json:"totalMinted"`
+	UpgradeEpochNum        *hexutil.Big `json:"upgradeEpochNum"`
+	EpochNum               *hexutil.Big `json:"epochNum"`
+	BlockHash              common.Hash  `json:"blockHash"`
+	BlockNumber            *hexutil.Big `json:"blockNumber"`
+}
+
+func (s *BlockChainAPI) GetTokenSupply(ctx context.Context, epochNr rpc.EpochNumber) (*tokenSupply, error) {
+	engine, ok := s.b.Engine().(*XDPoS.XDPoS)
+	if !ok {
+		return nil, errors.New("Undefined XDPoS consensus engine")
```

Comment:
error string should started with lowercase,  use `undefined` instead of `Undefined`